### PR TITLE
when testing downstream use the correct release branch

### PR DESCRIPTION
**Pull request by dprotaso**
additionally, we re-merge upstream in case new changes appear in HEAD

Fixes: https://github.com/knative/pkg/pull/2265

### CODE REVIEW

File: .github/workflows/knative-downstream.yaml [hunk 1]
```diff
@@ -80,18 +80,33 @@ jobs:
...
+
     - name: Install Dependencies
       run: |
         go get github.com/google/go-licenses
-    - name: Checkout Upstream
+
+    - name: Check out code
       uses: actions/checkout@v2
       with:
+        fetch-depth: 0
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:91]: this is the default, no need to set it to zero

File: .github/workflows/knative-downstream.yaml [hunk 2]
```diff
@@ -80,18 +80,33 @@ jobs:
       uses: actions/setup-go@v2
       with:
         go-version: 1.16.x
+
     - name: Install Dependencies
       run: |
         go get github.com/google/go-licenses
-    - name: Checkout Upstream
+
+    - name: Check out code
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:88]:
```suggestion
    - name: Checkout Upstream at ${{ github.base_ref }}
```

**Comment by dprotaso** [comment]
oh weird - something broke
/hold

**Comment by dprotaso** [comment]
looks like fetch-depth default is 1 and not 0 - so we'll probably want 0

**Comment by dprotaso** [comment]
@n3wscott PTAL

File: .github/workflows/knative-downstream.yaml [hunk 3]
```diff
@@ -80,19 +80,34 @@ jobs:
...
-    - name: Checkout Upstream
+
+    - name: Checkout upstream at ${{ github.base_ref }}
       uses: actions/checkout@v2
       with:
+        fetch-depth: 0
         path: ./src/knative.dev/${{ github.event.repository.name }}
-    - name: Checkout Downstream
+
+    - name: Merge upstream
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:94]: Why is using ref in the checkout not enough?

File: .github/workflows/knative-downstream.yaml [hunk 4]
```diff
@@ -80,19 +80,34 @@ jobs:
...
+      if: github.event_name == 'pull_request'
+      shell: bash
+      working-directory: ./src/knative.dev/${{ github.event.repository.name }}
+      run: |
+         git remote add upstream https://github.com/${{ github.repository }}.git
+         git fetch upstream ${{ github.base_ref }}
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:100]: What is baseref in the case of a pull request?

File: .github/workflows/knative-downstream.yaml [hunk 5]
```diff
@@ -80,19 +80,34 @@ jobs:
...
+         git pull --no-rebase --no-commit upstream ${{ github.base_ref }}
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:101]: I think what you are going to be doing is checking out the fork, and then pulling the forks branch that the pr is based on and what you are trying to do is checkout the original repo and try to merge the fork onto it? I am not sure this is needed, if the repo has broken the API for the downstream, it should have been reported, and there would be a conflict?

File: .github/workflows/knative-downstream.yaml [hunk 6]
```diff
@@ -80,19 +80,34 @@ jobs:
...
+
     - name: Install Dependencies
       run: |
         go get github.com/google/go-licenses
-    - name: Checkout Upstream
+
+    - name: Checkout upstream at ${{ github.base_ref }}
       uses: actions/checkout@v2
       with:
+        fetch-depth: 0
```
Comment (n3wscott) [inline .github/workflows/knative-downstream.yaml:91]: Remove the default?

**Comment by n3wscott** [comment]
I am not sure this is checking the HEAD repo out, but I could.be wrong. It seems it is merging the base branch back onto the pr from the fork repo? Please confirm.

File: .github/workflows/knative-downstream.yaml (hunk 3 above)
Comment (dprotaso) [reply .github/workflows/knative-downstream.yaml:94]: after a PR sits for a few days we want to retest with any changes `HEAD`

Comment (dprotaso) [reply .github/workflows/knative-downstream.yaml:94]: especially if there are failed checks - this what prow does as well

File: .github/workflows/knative-downstream.yaml (hunk 4 above)
Comment (dprotaso) [reply .github/workflows/knative-downstream.yaml:100]: Whatever the target branch is - ie. `main` or `release-*`

File: .github/workflows/knative-downstream.yaml (hunk 6 above)
Comment (dprotaso) [reply .github/workflows/knative-downstream.yaml:91]: the default is `1` not 0

File: .github/workflows/knative-downstream.yaml (hunk 5 above)
Comment (dprotaso) [reply .github/workflows/knative-downstream.yaml:101]: This pulls in the latest commits from `HEAD` into the PRs changes. We already do this prior to running unit tests etc.

**Comment by dprotaso** [comment]
> It seems it is merging the base branch back onto the pr from the fork repo? Please confirm.

Upstream is here:

`git remote add upstream https://github.com/${{ github.repository }}.git`


repository is where the event occurred - https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads#pull_request

**Comment by dprotaso** [comment]
bump @n3wscott

**Review by n3wscott — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**